quarkchain/cluster/jsonrpc.py

```diff
@@ -1361,3 +1363,73 @@ def get_data_default(key, decoder, default=None):
         else:  # estimate gas
             res = await self.master.estimate_gas(tx, sender_address)
             return quantity_encoder(res) if res is not None else None
+
+
+# JSONRPC Websocket Server
+class JSONRPCWSServer:
+    @classmethod
+    def start_public_server(cls, env, slave_server):
+        server = cls(
+            env,
+            slave_server,
+            env.cluster_config.JSON_RPC_WS_PORT,
+            env.cluster_config.JSON_RPC_WS_HOST,
+            public_methods,
+        )
+        server.start()
+        return server
+
+    def __init__(
+        self, env, slave_server: SlaveServer, port, host, methods: AsyncMethods
+    ):
+        self.loop = asyncio.get_event_loop()
+        self.port = port
+        self.host = host
+        self.env = env
+        self.master = slave_server
+        self.counters = dict()
+
+        # Bind RPC handler functions to this instance
+        self.handlers = AsyncMethods()
+        for rpc_name in methods:
+            func = methods[rpc_name]
+            self.handlers[rpc_name] = func.__get__(self, self.__class__)
+
+    async def __handle(self, websocket, path):
+        request = await websocket.recv()
+        Logger.info(request)
+
+        d = dict()
+        try:
+            d = json.loads(request)
+        except Exception:
+            pass
+        method = d.get("method", "null")
+        if method in self.counters:
+            self.counters[method] += 1
+        else:
+            self.counters[method] = 1
+
+        response = await self.handlers.dispatch(request)
+        if "error" in response:
+            Logger.error(response)
+        if not response.is_notification:
+            await websocket.send(str(response))
+        return response
+
+    def start(self):
+        start_server = websockets.serve(self.__handle, "localhost", 5000)
+        self.loop.run_until_complete(start_server)
+        self.loop.run_forever()
```

Comment:
in production you can't start a server like this. should follow http jsonrpc server